src/append_only_data.rs

```diff
@@ -688,6 +680,167 @@ where
     }
 }
 
+macro_rules! check_perm {
+    ($data: ident, $requester: ident, $request: ident) => {
+        if $data
+            .inner
+            .owners
```

Comment:
It would be more elegant here and future-proof to use the provided AData interface instead of accessing the internals directly. Something like this:

```rust
if $data.fetch_owner_at_index(data.owners_index()-1)...
```

Maybe you could even add a `last_owner` method and use that instead, and rename the `last` method to `last_data`.

Same for `inner.permissions.last()` below :)